### DISCUSSION TITLE
Looking for alternate method of MD5() which is deprecated in OpenSSL 3.0

### DISCUSSION THREAD

**Discussion by debananda-pal** [openssl/openssl#23493] [Q&A] [answered]
Hi All,

We are getting error while upgrading one of our component infra-etl, where below MD5 function is used :
#  ifndef OPENSSL_NO_DEPRECATED_3_0
OSSL_DEPRECATEDIN_3_0 int MD5_Init(MD5_CTX *c);
OSSL_DEPRECATEDIN_3_0 int MD5_Update(MD5_CTX *c, const void *data, size_t len);
OSSL_DEPRECATEDIN_3_0 int MD5_Final(unsigned char *md, MD5_CTX *c);
**OSSL_DEPRECATEDIN_3_0 unsigned char *MD5(const unsigned char *d, size_t n,
                                         unsigned char *md);**
OSSL_DEPRECATEDIN_3_0 void MD5_Transform(MD5_CTX *c, const unsigned char *b);
#  endif

I am getting below error :

../inc/etl_types.h:183:12: error: 'unsigned char* MD5(const unsigned char*, size_t, unsigned char*)' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  183 |         MD5((unsigned char*)target.c_str(), target.size(), digest);
      |         ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../inc/etl_types.h:12:
/usr/include/openssl/md5.h:52:38: note: declared here
   52 | OSSL_DEPRECATEDIN_3_0 unsigned char *MD5(const unsigned char *d, size_t n,
      |                                      ^~~
cc1plus: all warnings being treated as errors

Please let me know if there is any alternate method or other approach to be taken.

Regards,
D Pal

**Comment by levitte** [comment, marked as answer]
The absolutely quickest replacement is this:

``` C
EVP_Q_digest(NULL, "MD5", NULL, target.c_str(), target.size(), digest, NULL);
```

Don't forget to check the return value, which is 1 for success and 0 for failure (with an OpenSSL error record raised).  You may also want to have a look at the documentation for its arguments (all the stuff I left NULL).

**Reply by setop** [reply]
I get undefined reference to `EVP_Q_digest'

**Reply by levitte** [reply]
In what openssl version?

**Comment by debananda-pal** [comment]
Thank you levitte.
